### PR TITLE
[Feat]: BETWEEN for Filters

### PR DESCRIPTION
**Pull request by johannesschobel**
Dear @doug-martin ,

this PR adds the feature to filter a parameter with `between`. Currently, the syntax is as follows:

```gql
{
  todos(filter: { 
    priority: { between: { lower: 1, upper: 10 } } 
  } ) {
    edges {
      node { id }
    }
  }
}
```

I did my best to figure out, how this filter stuff works - and i think i understood it. Note, that i only did the changes for `TypeORM` as i am not familiar with `Sequelize`.

Finally, this PR fixes some typos in the context of this PR:
```diff
// note the "i"
- comparision
+ comparison
```

Feel free to add comments, suggestions, changes, ...
Thanks a lot for your time reviewing this PR!

All the best

### CODE REVIEW

**Comment by doug-martin** [comment]
@johannesschobel thanks for the PR! 

The sequleize support should be fairly straight forward, take a look at https://sequelize.org/v5/manual/querying.html#operators and https://github.com/doug-martin/nestjs-query/blob/master/packages/query-sequelize/src/query/sql-comparison.builder.ts#L14. The SQLComparison builder is a mapping between sequelize operators and nestjs-query operation. Shouldn't be too hard to add :) 

Could you also add the not between for both typeorm and sequelize, I've tried to keep comparisons and their inverted forms.

Do you know about dialect support for the between operator. I've ran into issues before with https://github.com/doug-martin/goqu and https://github.com/c2fo/patio but I can't recall what the exact issue was.

**Comment by doug-martin** [comment]
Have given much thought to the DSL? I've seen both the `upper` and `lower` or `[lower, upper]` (sequelize) I'm partial to the latter but I'm not sure how well that plays with graphql. 

I know this is a little forward thinking but along the same lines as dialect support should we just create `(val > col AND val < col)` for ORMs that dont support between natively?

**Comment by johannesschobel** [comment]
Dear @doug-martin ,

thank you very much for your feedback on my PR.

-----

What do you think with the "thoughts on the DSL"? Do you mean "my" DSL when submitting the `filter` on the `GraphQL` endpoint, like the

```gql
{
  filter: {
    FIELD: { between: { lower: 1, upper: 10 } }
  }
}
```

or what do you mean?

I would like to have it properly "typed" (i.e., with `lower` and `upper`), instead of just submitting an array with 2 fields, because you cannot properly enforce these rules. I would really love to keep this "nested" structured i have proposed in this PR.

-----

Of course i can add a `notBetween` operation as well. I can take a look at this later on..

-----

Regarding "Dialect Support":

Before actually writing the PR, i checked the docs on `BETWEEN` for the following major databases:
- PostgreSQL
- MySQL
- MariaDB
- SQLite

in order to see differences - but was not able to find any. Therefore, I assume, that all major DBs follow the same syntax and semantics.

I am not sure, if there are actually some "known databases" that **do not** support `between` operations. I mean, even `SQLite` supports `BETWEEN` :laughing: 

-----

I will try to look at Sequelize later today.. maybe i can find a solution. Maybe i just need to add an `if` statement to the `build` method` :thinking:

**Comment by johannesschobel** [comment]
Dear @doug-martin ,

i just updated the PR and added the following new features:
- [x] `notBetween` operator for `typeorm`
- [x] `between` operator for `sequelize`
- [x] `notBetween` operator for `sequelize`
- [x] updated the default `sequelize` project to contain an `int` value (`todo.priority`) to test the `between`/`notBetween` feature
- [x] adapted `gql schema files` to reflect those changes and allow tests to pass

Can you please give feedback?
Thanks a lot for your time and help

File: packages/query-sequelize/__tests__/query/sql-comparison.builder.spec.ts [hunk 1]
```diff
@@ -1,18 +1,18 @@
 import { Op } from 'sequelize';
 import { TestEntity } from '../__fixtures__/test.entity';
-import { SQLComparisionBuilder } from '../../src/query';
+import { SQLComparisonBuilder } from '../../src/query';
 
-describe('SQLComparisionBuilder', (): void => {
-  const createSQLComparisionBuilder = () => new SQLComparisionBuilder<TestEntity>();
+describe('SQLComparisonBuilder', (): void => {
```
Comment (doug-martin) [inline packages/query-sequelize/__tests__/query/sql-comparison.builder.spec.ts:5]: Can you please add a test for `between` and `notBetween`?

File: packages/query-sequelize/src/query/sql-comparison.builder.ts [hunk 2]
```diff
@@ -48,6 +48,40 @@ export class SQLComparisionBuilder<Entity> {
...
     throw new Error(`unknown operator "${cmp}"`);
   }
+
+  private betweenComparisonSQL<F extends keyof Entity>(
+    col: string,
+    val: EntityComparisonField<Entity, F>,
+  ): WhereOptions {
+    // TODO: this is not very good - can we better type this?!
+    // eslint-disable-next-line
+    const value = val as any;
```
Comment (doug-martin) [inline packages/query-sequelize/src/query/sql-comparison.builder.ts:68]: Yeah this isnt the greatest :( We'll need to play with the type def above. For now I think the following should work.

```suggestion
const value = (val as unknown) as CommonFieldComparisonBetweenType<Entity[F]>;
```

File: packages/query-sequelize/src/query/sql-comparison.builder.ts [hunk 3]
```diff
@@ -48,6 +48,40 @@ export class SQLComparisionBuilder<Entity> {
...
+    };
+  }
+
+  private notBetweenComparisonSQL<F extends keyof Entity>(
+    col: string,
+    val: EntityComparisonField<Entity, F>,
+  ): WhereOptions {
+    // TODO: this is not very good - can we better type this?!
+    // eslint-disable-next-line
+    const value = val as any;
```
Comment (doug-martin) [inline packages/query-sequelize/src/query/sql-comparison.builder.ts:81]:
```suggestion
const value = (val as unknown) as CommonFieldComparisonBetweenType<Entity[F]>;
```

File: packages/query-typeorm/__tests__/query/sql-comparison.builder.spec.ts [hunk 4]
```diff
@@ -1,26 +1,26 @@
 import { TestEntity } from '../__fixtures__/test.entity';
-import { SQLComparisionBuilder } from '../../src/query';
+import { SQLComparisonBuilder } from '../../src/query';
 
-describe('SQLComparisionBuilder', (): void => {
-  const createSQLComparisionBuilder = () => new SQLComparisionBuilder<TestEntity>();
+describe('SQLComparisonBuilder', (): void => {
```
Comment (doug-martin) [inline packages/query-typeorm/__tests__/query/sql-comparison.builder.spec.ts:4]: Can you please add a test for `between` and `notBetween`?

File: packages/query-typeorm/src/query/sql-comparison.builder.ts [hunk 5]
```diff
@@ -139,4 +147,37 @@ export class SQLComparisionBuilder<Entity> {
       throw new Error(`Invalid in value expected a non-empty array got ${JSON.stringify(val)}`);
     }
   }
+
+  private betweenComparisonSQL<F extends keyof Entity>(col: string, val: EntityComparisonField<Entity, F>): CmpSQLType {
+    const { paramName } = this;
+    // TODO: this is not very good - can we better type this?!
+    // eslint-disable-next-line
+    const value = val as any;
```
Comment (doug-martin) [inline packages/query-typeorm/src/query/sql-comparison.builder.ts:155]:
```suggestion
const value = (val as unknown) as CommonFieldComparisonBetweenType<Entity[F]>;
```

File: packages/query-typeorm/src/query/sql-comparison.builder.ts [hunk 6]
```diff
@@ -139,4 +147,37 @@ export class SQLComparisionBuilder<Entity> {
...
+  }
+
+  private notBetweenComparisonSQL<F extends keyof Entity>(
+    col: string,
+    val: EntityComparisonField<Entity, F>,
+  ): CmpSQLType {
+    const { paramName } = this;
+    // TODO: this is not very good - can we better type this?!
+    // eslint-disable-next-line
+    const value = val as any;
```
Comment (doug-martin) [inline packages/query-typeorm/src/query/sql-comparison.builder.ts:173]:
```suggestion
const value = (val as unknown) as CommonFieldComparisonBetweenType<Entity[F]>;
```

**Comment by johannesschobel** [comment]
Dear @doug-martin ,
Thank you very much for your time reviewing my PR. Glad you like it.
I will take a look at your suggestion / comments tomorrow, as it's a public holiday today :)

Have a nice day and thanks a lot again..
Johannes

**Comment by johannesschobel** [comment]
Dear @doug-martin ,
i was able to add your suggested better typings for the `typeorm query builder` **but NOT** for the `sequelize builder`. Unfortunately, the `sequelize builder` was not able to properly cast it in order provide a `WhereOptions` return type. Don't know why this fails.

Furthermore, i was not able to write a test for the `between / notBetween` method, as the `build()` method strangely did not accept the object with `lower / upper`. Am i missing something here?

Thanks for your input

**Comment by doug-martin** [comment]
@johannesschobel I merged into a local branch and updated with my latest eslint changes, and added a few tests. Check out #222 

Thank you for all the work you put into this!

**Comment by doug-martin** [comment]
Published! `v0.11.6`

**Comment by johannesschobel** [comment]
Dear @doug-martin , thanks a lot for reviewing and merging my PR

**Pull request closed without merging**